src/irc_quotegrabs.lua

```diff
@@ -0,0 +1,148 @@
+local sql = require 'lsqlite3'
+
+local wl_ins = nil
+local wl_del = nil
+local wl_sel = nil
+
+local qg_ins = nil
+local qg_sel_nick = nil
+local qg_sel_id = nil
+local qg_sel_rand = nil
+local qg_search = nil
+local qg_del = nil
+
+local quotegrabs = {}
+
+-- table to keep track of each nick's last message
+quotegrabs.last_msgs = {}
+
+quotegrabs.whitelist_status = function (nick)
+    wl_sel:reset()
+    wl_sel:bind_names({ ['nick'] = nick })
+    for c in wl_sel:urows() do
+        return c > 0
+    end
+end
+
+quotegrabs.whitelist_add = function (nick)
+    wl_ins:reset()
+    wl_ins:bind_names({ ['nick'] = nick })
+    local res = wl_ins:step()
+    return (res == 101 and 'Tada!' or db:errmsg())
+end
+
+quotegrabs.whitelist_del = function (nick)
+    wl_del:reset()
+    wl_del:bind_names({ ['nick'] = nick })
+    local res = wl_del:step()
+    return (res == 101 and 'Tada!' or db:errmsg())
+end
+
+quotegrabs.add = function (nick, msg)
+    qg_ins:reset()
+    qg_ins:bind_names({ ['nick'] = nick, ['message'] = msg })
+    local res = qg_ins:step()
+    return (res == 101 and 'Tada!' or db:errmsg())
+end
+
+quotegrabs.quote_nick = function (nick)
+    nick = (not nick or nick == '') and '%' or ('%' .. nick .. '%')
+    qg_sel_nick:reset()
+    qg_sel_nick:bind_names({ ['nick'] = nick })
+    for id, nick, msg in qg_sel_nick:urows() do
+        return id, nick, msg
+    end
+end
+
+quotegrabs.quote_id = function (id)
+    qg_sel_id:reset()
+    qg_sel_id:bind_names({ ['id'] = id })
+    for id, nick, msg in qg_sel_id:urows() do
+        return id, nick, msg
+    end
+end
+
+quotegrabs.quote_rand = function (nick)
+    nick = (not nick or nick == '') and '%' or ('%' .. nick .. '%')
+    qg_sel_rand:reset()
+    qg_sel_rand:bind_names({ ['nick'] = nick })
+    for id, nick, msg in qg_sel_rand:urows() do
+        return id, nick, msg
+    end
+end
+
+function truncate_string(str, len)
+    if str:len() > len then
+        str = str:sub(1, len) .. '...'
+    end
+    return str
+end
+
+quotegrabs.quote_search = function (search)
+    search = (not search or search == '') and '%' or ('%' .. search .. '%')
+    qg_search:reset()
+    qg_search:bind_names({ ['search'] = search })
+
+    local list = ''
+    for id, nick, msg in qg_search:urows() do
+        list = ('#%s: %s: %s, %s'):format(id, nick, truncate_string(msg, 40), list)
+    end
+
+    return list
+end
+
+quotegrabs.delete = function (id)
+    qg_del:reset()
+    qg_del:bind_names({ ['id'] = id })
+    local res = qg_del:step()
+    return (res == 101 and 'Tada!' or db:errmsg())
+end
+
+quotegrabs.dbinit = function ()
+    local quoteinit = [=[
+      create table if not exists quotegrabs (
+          id integer primary key not null,
+          nick text not null,
+          message text not null
+      );
+    ]=]
+
+    local whitelistinit = [=[
+      create table if not exists qg_whitelist (
+          nick text not null
+    );
```

Comment:
as long as we're making changes, this is a few characters left than it should be (it should be aligned with `create`).